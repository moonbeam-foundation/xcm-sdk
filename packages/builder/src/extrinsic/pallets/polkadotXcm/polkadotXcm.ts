import { ExtrinsicConfig } from '../../../types/substrate/ExtrinsicConfig';
import type { ExtrinsicConfigBuilder } from '../../ExtrinsicBuilder.interfaces';
import {
  getExtrinsicAccount,
  getExtrinsicArgumentVersion,
  normalizeConcrete,
  normalizeX1,
} from '../../ExtrinsicBuilder.utils';
import {
  getPolkadotXcmExtrinsicArgs,
  shouldFeeAssetPrecedeAsset,
} from './polkadotXcm.util';

const pallet = 'polkadotXcm';

export function polkadotXcm() {
  return {
    limitedReserveTransferAssets: () => {
      const func = 'limitedReserveTransferAssets';

      return {
        here: (): ExtrinsicConfigBuilder => ({
          build: (params) =>
            new ExtrinsicConfig({
              module: pallet,
              func,
              getArgs: (extrinsicFunction) => {
                const version = getExtrinsicArgumentVersion(extrinsicFunction);
                return getPolkadotXcmExtrinsicArgs({
                  ...params,
                  func: extrinsicFunction,
                  asset: [
                    {
                      id: normalizeConcrete(version, {
                        parents: 0,
                        interior: 'Here',
                      }),
                      fun: {
                        Fungible: params.asset.amount,
                      },
                    },
                  ],
                });
              },
            }),
        }),
        X1: (): ExtrinsicConfigBuilder => ({
          build: (params) =>
            new ExtrinsicConfig({
              module: pallet,
              func,
              getArgs: (extrinsicFunction) => {
                const version = getExtrinsicArgumentVersion(extrinsicFunction);

                return getPolkadotXcmExtrinsicArgs({
                  ...params,
                  func: extrinsicFunction,
                  asset: [
                    {
                      id: normalizeConcrete(
                        version,
                        normalizeX1(version, {
                          parents: 0,
                          interior: {
                            X1: {
                              PalletInstance:
                                params.asset.getAssetPalletInstance(),
                            },
                          },
                        }),
                      ),
                      fun: {
                        Fungible: params.asset.amount,
                      },
                    },
                  ],
                });
              },
            }),
        }),
        X2: (): ExtrinsicConfigBuilder => ({
          build: (params) =>
            new ExtrinsicConfig({
              module: pallet,
              func,
              getArgs: (extrinsicFunction) => {
                const isAssetDifferent = !params.asset.isSame(params.fee);
                const version = getExtrinsicArgumentVersion(extrinsicFunction);

                const assets = [
                  {
                    id: normalizeConcrete(version, {
                      parents: 0,
                      interior: {
                        X2: [
                          {
                            PalletInstance:
                              params.asset.getAssetPalletInstance(),
                          },
                          {
                            GeneralIndex: params.asset.getAssetId(),
                          },
                        ],
                      },
                    }),
                    fun: {
                      Fungible: params.asset.amount,
                    },
                  },
                ];

                const shouldFeeAssetPrecede =
                  shouldFeeAssetPrecedeAsset(params);

                if (isAssetDifferent) {
                  const feeAsset = {
                    id: normalizeConcrete(version, {
                      parents: 0,
                      interior: {
                        X2: [
                          {
                            PalletInstance: params.fee.getAssetPalletInstance(),
                          },
                          {
                            GeneralIndex: params.fee.getAssetId(),
                          },
                        ],
                      },
                    }),
                    fun: {
                      Fungible: params.fee.amount,
                    },
                  };

                  if (shouldFeeAssetPrecede) {
                    assets.unshift(feeAsset);
                  } else {
                    assets.push(feeAsset);
                  }
                }

                return getPolkadotXcmExtrinsicArgs({
                  ...params,
                  func: extrinsicFunction,
                  asset: assets,
                  feeIndex: isAssetDifferent && !shouldFeeAssetPrecede ? 1 : 0,
                });
              },
            }),
        }),
      };
    },
    limitedReserveWithdrawAssets: () => {
      const func = 'limitedReserveWithdrawAssets';

      return {
        X2: (): ExtrinsicConfigBuilder => ({
          build: (params) =>
            new ExtrinsicConfig({
              module: pallet,
              func,
              getArgs: (extrinsicFunction) => {
                const version = getExtrinsicArgumentVersion(extrinsicFunction);

                return getPolkadotXcmExtrinsicArgs({
                  ...params,
                  func: extrinsicFunction,
                  asset: [
                    {
                      id: normalizeConcrete(version, {
                        parents: 1,
                        interior: {
                          X2: [
                            {
                              Parachain: params.destination.parachainId,
                            },
                            {
                              PalletInstance:
                                params.asset.getAssetPalletInstance(),
                            },
                          ],
                        },
                      }),
                      fun: {
                        Fungible: params.asset.amount,
                      },
                    },
                  ],
                });
              },
            }),
        }),
      };
    },
    transferAssets: () => {
      const func = 'transferAssets';

      return {
        here: (): ExtrinsicConfigBuilder => ({
          build: (params) =>
            new ExtrinsicConfig({
              module: pallet,
              func,
              getArgs: (extrinsicFunction) => {
                const version = getExtrinsicArgumentVersion(extrinsicFunction);

                return getPolkadotXcmExtrinsicArgs({
                  ...params,
                  func: extrinsicFunction,
                  asset: [
                    {
                      id: normalizeConcrete(version, {
                        parents: 1,
                        interior: 'Here',
                      }),
                      fun: {
                        Fungible: params.asset.amount,
                      },
                    },
                  ],
                });
              },
            }),
        }),
        X2AndFeeHere: (): ExtrinsicConfigBuilder => ({
          build: (params) =>
            new ExtrinsicConfig({
              module: pallet,
              func,
              getArgs: (extrinsicFunction) => {
                const version = getExtrinsicArgumentVersion(extrinsicFunction);

                return getPolkadotXcmExtrinsicArgs({
                  ...params,
                  func: extrinsicFunction,
                  asset: [
                    {
                      id: normalizeConcrete(version, {
                        parents: 0,
                        interior: {
                          X2: [
                            {
                              PalletInstance:
                                params.asset.getAssetPalletInstance(),
                            },
                            {
                              GeneralIndex: params.asset.getAssetId(),
                            },
                          ],
                        },
                      }),
                      fun: {
                        Fungible: params.asset.amount,
                      },
                    },
                    // Fee Asset
                    {
                      id: normalizeConcrete(version, {
                        parents: 1,
                        interior: 'Here',
                      }),
                      fun: {
                        Fungible: params.fee.amount,
                      },
                    },
                  ],
                  feeIndex: 1,
                });
              },
            }),
        }),
      };
    },
    transferAssetsUsingTypeAndThen: () => {
      const func = 'transferAssetsUsingTypeAndThen';
      return {
        // TODO we could pass a parameter globalConsensus in the chain asset if we need a different one
        globalConsensusEthereum: (): ExtrinsicConfigBuilder => ({
          build: (params) =>
            new ExtrinsicConfig({
              module: pallet,
              func,

              getArgs: (extrinsicFunction) => {
                if (!params.asset.address) {
                  throw new Error(
                    'Asset address is needed for transferAssetsUsingTypeAndThen.globalConsensus function',
                  );
                }

                const version = getExtrinsicArgumentVersion(extrinsicFunction);
                return [
                  {
                    [version]: {
                      parents: 1,
                      interior: {
                        X1: [
                          {
                            Parachain: params.destination.parachainId,
                          },
                        ],
                      },
                    },
                  },
                  {
                    [version]: [
                      {
                        id: normalizeConcrete(version, {
                          parents: 2,
                          interior: {
                            X2: [
                              { globalConsensus: { Ethereum: { ChainId: 1 } } },
                              getExtrinsicAccount(params.asset.address),
                            ],
                          },
                        }),
                        fun: { fungible: params.asset.amount },
                      },
                    ],
                  },
                  'LocalReserve',
                  {
                    [version]: normalizeConcrete(version, {
                      parents: 2,
                      interior: {
                        X2: [
                          { globalConsensus: { Ethereum: { ChainId: 1 } } },
                          getExtrinsicAccount(params.asset.address),
                        ],
                      },
                    }),
                  },
                  'LocalReserve',
                  {
                    [version]: [
                      {
                        DepositAsset: {
                          assets: {
                            Wild: { AllCounted: 1 },
                          },
                          beneficiary: normalizeX1(version, {
                            parents: 0,
                            interior: {
                              X1: getExtrinsicAccount(
                                params.destinationAddress,
                              ),
                            },
                          }),
                        },
                      },
                    ],
                  },
                  'Unlimited',
                ];
              },
            }),
        }),
      };
    },
  };
}
