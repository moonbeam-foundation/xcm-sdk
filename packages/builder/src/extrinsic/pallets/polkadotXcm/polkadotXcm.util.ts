import type { SubmittableExtrinsicFunction } from '@polkadot/api/types';
import type { BuilderParams } from '../../../builder.interfaces';
import type { Parents } from '../../ExtrinsicBuilder.interfaces';
import {
  getExtrinsicAccount,
  getExtrinsicArgumentVersion,
  normalizeX1,
} from '../../ExtrinsicBuilder.utils';

export interface GetExtrinsicParams extends BuilderParams {
  // biome-ignore lint/suspicious/noExplicitAny: not sure how to fix this
  asset: any;
  func?: SubmittableExtrinsicFunction<'promise'>;
  parents?: Parents;
  feeIndex?: number;
}

export function getPolkadotXcmExtrinsicArgs({
  asset,
  destinationAddress,
  destination,
  func,
  parents = 1,
  feeIndex = 0,
}: GetExtrinsicParams) {
  const version = getExtrinsicArgumentVersion(func);

  return [
    {
      [version]: normalizeX1(version, {
        parents,
        interior: {
          X1: {
            Parachain: destination.parachainId,
          },
        },
      }),
    },
    {
      [version]: normalizeX1(version, {
        parents: 0,
        interior: {
          X1: getExtrinsicAccount(destinationAddress),
        },
      }),
    },
    {
      [version]: asset,
    },
    feeIndex,
    'Unlimited',
  ];
}

export function shouldFeeAssetPrecedeAsset({
  asset,
  fee,
}: BuilderParams): boolean {
  const assetIdNumber = Number(asset.getAssetId());
  const feeAssetIdNumber = Number(fee.getAssetId());

  if (Number.isNaN(assetIdNumber) || Number.isNaN(feeAssetIdNumber)) {
    return false;
  }

  return assetIdNumber > feeAssetIdNumber;
}
