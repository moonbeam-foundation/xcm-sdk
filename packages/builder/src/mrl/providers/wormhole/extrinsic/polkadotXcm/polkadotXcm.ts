import {
  type AnyParachain,
  AssetAmount,
  Parachain,
} from '@moonbeam-network/xcm-types';
import { getMultilocationDerivedAddresses } from '@moonbeam-network/xcm-utils';
import type { ApiPromise } from '@polkadot/api';
import type { SubmittableExtrinsic } from '@polkadot/api/types';
import type { ISubmittableResult } from '@polkadot/types/types';
import { ExtrinsicBuilder } from '../../../../../extrinsic/ExtrinsicBuilder';
import { ExtrinsicConfig } from '../../../../../types/substrate/ExtrinsicConfig';
import type {
  MrlBuilderParams,
  MrlConfigBuilder,
} from '../../../../MrlBuilder.interfaces';

// TODO: these have to come from the configs
const BUY_EXECUTION_FEE = 100_000_000_000_000_000n; // moonChainFee
export const CROSS_CHAIN_FEE = 100_000_000_000_000_000n; // fee for processing the xcm message in moon chain

export function polkadotXcm() {
  return {
    send: (): MrlConfigBuilder => ({
      build: ({
        asset,
        destination,
        destinationAddress,
        fee,
        isAutomatic,
        moonAsset,
        moonChain,
        moonApi,
        sendOnlyRemoteExecution,
        source,
        sourceAddress,
        sourceApi,
        transact,
      }) => {
        if (!destination.wh?.name) {
          throw new Error('Destination chain does not have a wormhole name');
        }

        if (!sourceApi) {
          throw new Error('Source API needs to be defined');
        }

        if (!Parachain.is(source)) {
          throw new Error('Source chain needs to be a parachain');
        }

        const { address20: computedOriginAccount } =
          getMultilocationDerivedAddresses({
            address: sourceAddress,
            paraId: source.parachainId,
            isParents: true,
          });

        const assetTransferTxs = getAssetTransferTxs({
          asset,
          computedOriginAccount,
          destination,
          destinationAddress,
          fee,
          isAutomatic,
          moonApi,
          moonAsset,
          moonChain,
          source,
          sourceAddress,
          sourceApi,
        });

        const send = buildSendExtrinsic({
          asset,
          destination,
          destinationAddress,
          computedOriginAccount,
          fee,
          isAutomatic,
          moonAsset,
          moonChain,
          moonApi,
          source,
          sourceAddress,
          sourceApi,
          transact,
        });

        const transactionsToSend = sendOnlyRemoteExecution
          ? [send]
          : [...assetTransferTxs, send];

        return new ExtrinsicConfig({
          module: 'utility',
          func: 'batchAll',
          getArgs: () => [transactionsToSend],
        });
      },
    }),
  };
}

interface HelperFunctionParams extends MrlBuilderParams {
  computedOriginAccount: string;
  source: AnyParachain;
  sourceApi: ApiPromise;
}

export function buildSendExtrinsic({
  computedOriginAccount,
  moonAsset,
  moonChain,
  sourceApi,
  transact,
}: HelperFunctionParams) {
  if (!transact) {
    throw new Error('Transact params are required');
  }

  return sourceApi.tx.polkadotXcm.send(
    {
      // TODO apply XCM versioning
      V3: {
        parents: 1,
        interior: { X1: { Parachain: moonChain.parachainId } },
      },
    },
    {
      V3: [
        {
          WithdrawAsset: [
            {
              id: {
                Concrete: {
                  parents: 0,
                  interior: {
                    X1: {
                      PalletInstance: moonAsset.getAssetPalletInstance(),
                    },
                  },
                },
              },
              fun: { Fungible: BUY_EXECUTION_FEE },
            },
          ],
        },
        {
          BuyExecution: {
            fees: {
              id: {
                Concrete: {
                  parents: 0,
                  interior: {
                    X1: {
                      PalletInstance: moonAsset.getAssetPalletInstance(),
                    },
                  },
                },
              },
              fun: { Fungible: BUY_EXECUTION_FEE },
            },
            weightLimit: 'Unlimited',
          },
        },
        {
          Transact: {
            originKind: 'SovereignAccount',
            requireWeightAtMost: {
              refTime: transact.txWeight.refTime,
              proofSize: transact.txWeight.proofSize,
            },
            call: {
              encoded: transact.call,
            },
          },
        },
        {
          SetAppendix: [
            { RefundSurplus: {} },
            {
              DepositAsset: {
                assets: { Wild: { AllCounted: 1 } },
                beneficiary: {
                  parents: 0,
                  interior: {
                    X1: { AccountKey20: { key: computedOriginAccount } },
                  },
                },
              },
            },
          ],
        },
      ],
    },
  );
}

function getAssetTransferTxs({
  asset,
  computedOriginAccount,
  fee,
  moonApi,
  moonAsset,
  moonChain,
  source,
  sourceAddress,
  sourceApi,
}: HelperFunctionParams): SubmittableExtrinsic<
  'promise',
  ISubmittableResult
>[] {
  const { transfer, transferMulticurrencies } = sourceApi.tx.xTokens;
  const transferBuilder = ExtrinsicBuilder().xTokens().transfer();
  const transferMulticurrenciesBuilder = ExtrinsicBuilder()
    .xTokens()
    .transferMultiCurrencies();
  /**
   * TODO here we should compare the asset with the cross chain fee asset.
   * For example, FTM cannot pay for fees in Moonbase while AGNG can, so for FTM we have to send a transferMulticurrencies
   * This "if" is a workaround, change when we implement properly the concept of cross-chain fee (different from moonChainFee)
   */
  if (asset.isSame(fee)) {
    const assetTransferTx = transfer(
      ...transferBuilder
        .build({
          asset: asset.copyWith({
            // TODO for the moment this is only applicable to peaq, AGNG pays for fees and we have to include the cross chain fee
            // for this we have to add a new concept in the config (Cross Chain Fee), and get the value from there
            amount: asset.amount + 10n * CROSS_CHAIN_FEE,
          }),
          destination: moonChain,
          destinationAddress: computedOriginAccount,
          destinationApi: moonApi,
          fee,
          source: source,
          sourceAddress,
          sourceApi,
        })
        .getArgs(transfer),
    );
    const feeAssetTransferTx = transfer(
      ...transferBuilder
        .build({
          asset: AssetAmount.fromChainAsset(source.getChainAsset(moonAsset), {
            amount: CROSS_CHAIN_FEE + BUY_EXECUTION_FEE,
          }),
          destination: moonChain,
          destinationAddress: computedOriginAccount,
          destinationApi: moonApi,
          fee,
          source: source,
          sourceAddress,
          sourceApi,
        })
        .getArgs(transfer),
    );
    return [assetTransferTx, feeAssetTransferTx];
  }
  const multiCurrenciesTransferTx = transferMulticurrencies(
    ...transferMulticurrenciesBuilder
      .build({
        asset,
        destination: moonChain,
        destinationAddress: computedOriginAccount,
        destinationApi: moonApi,
        fee: AssetAmount.fromChainAsset(source.getChainAsset(moonAsset), {
          amount: CROSS_CHAIN_FEE + BUY_EXECUTION_FEE,
        }),
        source: source as AnyParachain,
        sourceAddress,
        sourceApi,
      })
      .getArgs(transferMulticurrencies),
  );
  return [multiCurrenciesTransferTx];
}
