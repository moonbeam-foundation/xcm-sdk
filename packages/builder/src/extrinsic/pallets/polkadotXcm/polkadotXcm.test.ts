import { describe, expect, it } from 'vitest';

import {
  buildParachainParamsMock,
  buildParamsMock,
  buildParamsSameAssetMock,
} from '../../../../fixtures';
import { polkadotXcm } from './polkadotXcm';

describe('polkadotXcm', () => {
  describe('limitedReserveTransferAssets', () => {
    describe('here', () => {
      const extrinsic = polkadotXcm()
        .limitedReserveTransferAssets()
        .here()
        .build(buildParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });

    describe('x1', () => {
      const extrinsic = polkadotXcm()
        .limitedReserveTransferAssets()
        .X1()
        .build(buildParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });

    describe('x2', () => {
      describe('the same asset', () => {
        const extrinsic = polkadotXcm()
          .limitedReserveTransferAssets()
          .X2()
          .build(buildParamsSameAssetMock);

        it('should be correct config', () => {
          expect(extrinsic).toMatchSnapshot();
        });

        it('should get correct arguments', () => {
          expect(extrinsic.getArgs()).toMatchSnapshot();
        });
      });

      describe('different assets', () => {
        const extrinsic = polkadotXcm()
          .limitedReserveTransferAssets()
          .X2()
          .build(buildParamsMock);

        it('should be correct config', () => {
          expect(extrinsic).toMatchSnapshot();
        });

        it('should get correct arguments', () => {
          expect(extrinsic.getArgs()).toMatchSnapshot();
        });
      });
    });

    describe('parachain', () => {
      const extrinsic = polkadotXcm()
        .limitedReserveTransferAssets()
        .X2()
        .build(buildParachainParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });
  });

  describe('limitedReserveWithdrawAssets', () => {
    describe('x2', () => {
      const extrinsic = polkadotXcm()
        .limitedReserveWithdrawAssets()
        .X2()
        .build(buildParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });

    describe('parachain', () => {
      const extrinsic = polkadotXcm()
        .limitedReserveWithdrawAssets()
        .X2()
        .build(buildParachainParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });
  });

  describe('transferAssets', () => {
    describe('here', () => {
      const extrinsic = polkadotXcm()
        .transferAssets()
        .here()
        .build(buildParachainParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });
  });

  describe('transferAssets', () => {
    describe('X2AndFeeHere', () => {
      const extrinsic = polkadotXcm()
        .transferAssets()
        .X2AndFeeHere()
        .build(buildParachainParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });
  });

  describe('transferAssetsUsingTypeAndThen', () => {
    describe('globalConsensusEthereum', () => {
      const extrinsic = polkadotXcm()
        .transferAssetsUsingTypeAndThen()
        .globalConsensusEthereum()
        .build(buildParachainParamsMock);

      it('should be correct config', () => {
        expect(extrinsic).toMatchSnapshot();
      });

      it('should get correct arguments', () => {
        expect(extrinsic.getArgs()).toMatchSnapshot();
      });
    });
  });
});
