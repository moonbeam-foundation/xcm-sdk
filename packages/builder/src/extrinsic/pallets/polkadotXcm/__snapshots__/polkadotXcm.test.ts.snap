// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`polkadotXcm > limitedReserveTransferAssets > here > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "limitedReserveTransferAssets",
  "getArgs": [Function],
  "module": "polkadotXcm",
}
`;

exports[`polkadotXcm > limitedReserveTransferAssets > here > should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 1000,
        },
      },
      "parents": 1,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": "Here",
            "parents": 0,
          },
        },
      },
    ],
  },
  0,
  "Unlimited",
]
`;

exports[`polkadotXcm > limitedReserveTransferAssets > parachain > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "limitedReserveTransferAssets",
  "getArgs": [Function],
  "module": "polkadotXcm",
}
`;

exports[`polkadotXcm > limitedReserveTransferAssets > parachain > should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 2032,
        },
      },
      "parents": 1,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountId32": {
            "id": "0x18e6c7ad45fbab18778710642072fdd057470b76d4d568b44dea6accd5b0c423",
            "network": null,
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X2": [
                {
                  "PalletInstance": 10,
                },
                {
                  "GeneralIndex": "USDT",
                },
              ],
            },
            "parents": 0,
          },
        },
      },
      {
        "fun": {
          "Fungible": 5000000000n,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X2": [
                {
                  "PalletInstance": 18,
                },
                {
                  "GeneralIndex": "RMRK",
                },
              ],
            },
            "parents": 0,
          },
        },
      },
    ],
  },
  1,
  "Unlimited",
]
`;

exports[`polkadotXcm > limitedReserveTransferAssets > x1 > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "limitedReserveTransferAssets",
  "getArgs": [Function],
  "module": "polkadotXcm",
}
`;

exports[`polkadotXcm > limitedReserveTransferAssets > x1 > should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 1000,
        },
      },
      "parents": 1,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X1": {
                "PalletInstance": 10,
              },
            },
            "parents": 0,
          },
        },
      },
    ],
  },
  0,
  "Unlimited",
]
`;

exports[`polkadotXcm > limitedReserveTransferAssets > x2 > different assets > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "limitedReserveTransferAssets",
  "getArgs": [Function],
  "module": "polkadotXcm",
}
`;

exports[`polkadotXcm > limitedReserveTransferAssets > x2 > different assets > should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 1000,
        },
      },
      "parents": 1,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X2": [
                {
                  "PalletInstance": 10,
                },
                {
                  "GeneralIndex": "USDT",
                },
              ],
            },
            "parents": 0,
          },
        },
      },
      {
        "fun": {
          "Fungible": 5000000000n,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X2": [
                {
                  "PalletInstance": 18,
                },
                {
                  "GeneralIndex": "RMRK",
                },
              ],
            },
            "parents": 0,
          },
        },
      },
    ],
  },
  1,
  "Unlimited",
]
`;

exports[`polkadotXcm > limitedReserveTransferAssets > x2 > the same asset > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "limitedReserveTransferAssets",
  "getArgs": [Function],
  "module": "polkadotXcm",
}
`;

exports[`polkadotXcm > limitedReserveTransferAssets > x2 > the same asset > should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 1000,
        },
      },
      "parents": 1,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X2": [
                {
                  "PalletInstance": 10,
                },
                {
                  "GeneralIndex": "USDT",
                },
              ],
            },
            "parents": 0,
          },
        },
      },
    ],
  },
  0,
  "Unlimited",
]
`;

exports[`polkadotXcm > limitedReserveWithdrawAssets > parachain > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "limitedReserveWithdrawAssets",
  "getArgs": [Function],
  "module": "polkadotXcm",
}
`;

exports[`polkadotXcm > limitedReserveWithdrawAssets > parachain > should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 2032,
        },
      },
      "parents": 1,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountId32": {
            "id": "0x18e6c7ad45fbab18778710642072fdd057470b76d4d568b44dea6accd5b0c423",
            "network": null,
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X2": [
                {
                  "Parachain": 2032,
                },
                {
                  "PalletInstance": 10,
                },
              ],
            },
            "parents": 1,
          },
        },
      },
    ],
  },
  0,
  "Unlimited",
]
`;

exports[`polkadotXcm > limitedReserveWithdrawAssets > x2 > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "limitedReserveWithdrawAssets",
  "getArgs": [Function],
  "module": "polkadotXcm",
}
`;

exports[`polkadotXcm > limitedReserveWithdrawAssets > x2 > should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 1000,
        },
      },
      "parents": 1,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountKey20": {
            "key": "0xeF46c7649270C912704fB09B75097f6E32208b85",
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X2": [
                {
                  "Parachain": 1000,
                },
                {
                  "PalletInstance": 10,
                },
              ],
            },
            "parents": 1,
          },
        },
      },
    ],
  },
  0,
  "Unlimited",
]
`;

exports[`polkadotXcm > transferAssets > X2AndFeeHere > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transferAssets",
  "getArgs": [Function],
  "module": "polkadotXcm",
}
`;

exports[`polkadotXcm > transferAssets > X2AndFeeHere > should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 2032,
        },
      },
      "parents": 1,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountId32": {
            "id": "0x18e6c7ad45fbab18778710642072fdd057470b76d4d568b44dea6accd5b0c423",
            "network": null,
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X2": [
                {
                  "PalletInstance": 10,
                },
                {
                  "GeneralIndex": "USDT",
                },
              ],
            },
            "parents": 0,
          },
        },
      },
      {
        "fun": {
          "Fungible": 5000000000n,
        },
        "id": {
          "Concrete": {
            "interior": "Here",
            "parents": 1,
          },
        },
      },
    ],
  },
  1,
  "Unlimited",
]
`;

exports[`polkadotXcm > transferAssets > here > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transferAssets",
  "getArgs": [Function],
  "module": "polkadotXcm",
}
`;

exports[`polkadotXcm > transferAssets > here > should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": {
          "Parachain": 2032,
        },
      },
      "parents": 1,
    },
  },
  {
    "V1": {
      "interior": {
        "X1": {
          "AccountId32": {
            "id": "0x18e6c7ad45fbab18778710642072fdd057470b76d4d568b44dea6accd5b0c423",
            "network": null,
          },
        },
      },
      "parents": 0,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "Fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": "Here",
            "parents": 1,
          },
        },
      },
    ],
  },
  0,
  "Unlimited",
]
`;

exports[`polkadotXcm > transferAssetsUsingTypeAndThen > globalConsensusEthereum > should be correct config 1`] = `
ExtrinsicConfig {
  "func": "transferAssetsUsingTypeAndThen",
  "getArgs": [Function],
  "module": "polkadotXcm",
}
`;

exports[`polkadotXcm > transferAssetsUsingTypeAndThen > globalConsensusEthereum > should get correct arguments 1`] = `
[
  {
    "V1": {
      "interior": {
        "X1": [
          {
            "Parachain": 2032,
          },
        ],
      },
      "parents": 1,
    },
  },
  {
    "V1": [
      {
        "fun": {
          "fungible": 99000000000n,
        },
        "id": {
          "Concrete": {
            "interior": {
              "X2": [
                {
                  "globalConsensus": {
                    "Ethereum": {
                      "ChainId": 1,
                    },
                  },
                },
                {
                  "AccountKey20": {
                    "key": "0x98891e5FD24Ef33A488A47101F65D212Ff6E650E",
                  },
                },
              ],
            },
            "parents": 2,
          },
        },
      },
    ],
  },
  "LocalReserve",
  {
    "V1": {
      "Concrete": {
        "interior": {
          "X2": [
            {
              "globalConsensus": {
                "Ethereum": {
                  "ChainId": 1,
                },
              },
            },
            {
              "AccountKey20": {
                "key": "0x98891e5FD24Ef33A488A47101F65D212Ff6E650E",
              },
            },
          ],
        },
        "parents": 2,
      },
    },
  },
  "LocalReserve",
  {
    "V1": [
      {
        "DepositAsset": {
          "assets": {
            "Wild": {
              "AllCounted": 1,
            },
          },
          "beneficiary": {
            "interior": {
              "X1": {
                "AccountId32": {
                  "id": "0x18e6c7ad45fbab18778710642072fdd057470b76d4d568b44dea6accd5b0c423",
                  "network": null,
                },
              },
            },
            "parents": 0,
          },
        },
      },
    ],
  },
  "Unlimited",
]
`;
